### PR TITLE
Add some discussion of the security issues.

### PR DESCRIPTION
**Pull request by jyasskin**
* Outdated certificates and packages.
* That there may not be a single certificate chain.

@estark37

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -291,43 +316,50 @@ Content-Type: application/pkcs7-mime
...
+### Does the package supply a full chain of certificates to a known CA root?
 
-Not necessarily. To quote one of the Chrome security engineers, "because different devices have different sets of roots in their trust stores, it's not always the case that there is a single "correct" set of certificates to send that will work best for all clients. Instead, for compatibility, sites will sometimes send a set of certificates and rely on clients to dynamically fetch additional intermediates when needed". We suspect that to be an interesting issue since offline validation would need a full chain. Browsers (at least Chrome) are implementing automatic fetching of intermediate certificates but for this to work they have to be online. This will probably become a matter of best practices when creating the packaging format, possibly with validation by the packaging tool.
+Not necessarily. Different devices have different sets of roots in their trust
+stores, so there is not a single "correct" set of certificates to send that will
+work best for all clients. Instead, for compatibility, packages may need to
+include a set of certificates from which chains can be built to multiple roots,
```
Comment (jyasskin) [inline README.md]: @sleevi, @estark37 said I should ask you if packagers are going to cause problems by including the certificates needed to chain to the roots of multiple different clients. She referenced https://groups.google.com/a/chromium.org/d/msg/net-dev/H-ysp5UM_rk/TqoyjV1eDAAJ as an example of servers choosing to make the client download some intermediates, which won't work as well for offline verification.

Comment (sleevi) [inline README.md]: @jyasskin I'll read up on the full proposal, but it sounds like you're suggesting using CA-rooted certificates, possibly even TLS certificates, for the packaging. That's not ideal for a number of reasons - this included - so I'll take a scan through and see what alternatives might be worth suggesting.

Comment (jyasskin) [inline README.md]: Thanks. The main constraint that I think leads us to use CA-rooted certificates is that we need the packages to be trusted as coming from particular web origins. I'm happy with anything else that proves an origin to the displaying browser.

Comment (sleevi) [inline README.md]: Right, unfortunately, this introduces subtle-but-significant cross-protocol attacks when you re-use a TLS certificate to sign this data, unfortunately, so this is going to be quite difficult to get what you want.

Comment (dimich-g) [inline README.md]: @sleevi is this true even if the publisher of the package obtains a separate TLS certificate for signing packages? Sort of similar to how code-signing certificates are separate form TLS ones.

**Comment by lrosenthol** [comment]
This conversation has also taken place as part of #16. where its been pointed out that TLS-based certs won't work (as @sleevi says here).  Other types of certs - that aren't tied to domains, but instead of orgs or individuals - will need to be used.

But regardless, the set of valid and/or trusted roots needs to come from somewhere that may (or may not) be able to be mandated as part of the standardization.  And as part of that, you need to consider other standards in this area such as the [EUTL](https://ec.europa.eu/digital-single-market/en/eu-trusted-lists-trust-service-providers).

**Comment by sleevi** [comment]
@lrosenthol That's not an accurate statement - that is, that other types of certs _will_ need to be used, or more generally, that certificates not bound to domains should be used. And the EUTL should absolutely be ignored in this space - not even bothered with.

@dimich-g I think the question was worded incorrectly; that is, you don't want to get a "second" TLS certificate that you use for signing, because as far as design goes, you're setting people up to fail if they have two certificates, interchangable, but one should be used for TLS and one not. To reduce the risk of cross-protocol attacks, you want to ensure that Certificate A (which can/should/is bound to a domain) cannot be used in Protocol B, and that Certificate B (which can/should/is bound to a domain) should not be used in Protocol A.

I don't anticipate being able to provide a more thorough review of the cryptography in the next few days, but there are definitely red flags with what's currently specified - ranging from the discussion of Key Usage to these issues of cross-protocol confusion. The signature design also doesn't account for the ability to perform counter-signatures, which is generally seen as a necessary scheme in document/code signing solutions (in Authenticode, it's used for timestampping; in Chrome extensions, it's used to vet that it's a Google WebStore delivered extension; in Android APKs, it's used to comply with local laws).

**Comment by lrosenthol** [comment]
@sleevi 
>And the EUTL should absolutely be ignored in this space - not even bothered with.
Then you and I are going to have to agree to disagree.  The EUTL is a requirement for adoption of signatures within the URL for various business processes.  In the context of the use of this technology for publications, it would be necessary to be allow compatibility with it.  I am **not** suggesting mandating, just compatibility.

>doesn't account for the ability to perform counter-signatures
I don't believe that the current design prevents counter sigs, but I agree that it's not called out - and you are right, it should be as it is an important technology.  It's also going to be important for publications where the author may sign the original publication with the publisher adding theirs at the final release stage.

**Comment by jyasskin** [comment]
@sleevi At the moment, we don't have a specification, just a sketch. Would you prefer that we non-cryptographers come up with a draft specification that you can review, or would you rather we write down a list of requirements and wait for a real cryptographer to design the right system?

**Comment by sleevi** [comment]
@jyasskin Either way, knowing your goals clearly will help make sure whatever solution meets those goals - whether it's starting from a point in the draft where you're trying to accomplish those goals, or whether it's starting from scratch to build something to achieve those goals. Knowing, for example, whether you're exclusively targeting UAs or whether, as it seems suggested both here and the draft, you're imagining things like JS packaging or document delivery, is itself a useful aspect.

**Comment by lrosenthol** [comment]
@sleevi - there are at least three different groups that are looking to
potentially leverage this work
- Browsers (which I assume you meant when you said UA, but is just one
class of UA for the OWP)
- Application environments (eg. Node+Electron)
- Documents & Publications (eg. Web Publications)

This to me means a common model and file format.  However, I see no reason
why the types of certificates allowed/used couldn't be different in each
case.

**Comment by bmeck** [comment]
To clarify, I would expect Node and Electron to use Code Signing based x.509 certs rather than regular TLS certs. In particular, the origin of applications might not be tied to a specific WWW domain but an app store submission (like Android APK).

File: README.md [hunk 2]
```diff
@@ -8,9 +8,26 @@ Some new use cases for Web technology have motivated thinking about a multi-reso
 
 Local sharing is quite popular, especially in Emerging Markets countries, due to cost and limitations on cellular data and relatively spotty WiFi availability. It is typically done over local Bluetooth/WiFi, by either built-in OS features like [Android Beam](https://en.wikipedia.org/wiki/Android_Beam) or with popular 3-rd party apps, such as [ShareIt](https://play.google.com/store/apps/details?id=com.lenovo.anyshare.gps) or [Xender](https://play.google.com/store/apps/details?id=cn.xender)). Typically, the locally stored media files and apps (APK files for Android for example) are shared this way. Extending sharing to bundles of content and web apps (Progressive Web Apps in particular) opens up new possibilities, especially if combined with a form of signing the content. Cryptographic signing could make it possible to afford the shared content the treatment normally reserved for the origin that content is claiming to be in - by verifying that the content indeed was produced by the holder of the corresponding certificate.
 
+Local sharing tends to have persistent constraints on the amount of data
```
Comment (mrdewitt) [inline README.md]: I'm confused - local sharing doesn't typically have per-byte costs.  Storage and downloading do.

File: README.md [hunk 3]
```diff
@@ -8,9 +8,26 @@ Some new use cases for Web technology have motivated thinking about a multi-reso
...
+transferred: each byte costs money. This means the client may be stuck with an
+outdated version of a package for a significant amount of time, including that
+package's vulnerabilities. It may be feasible to periodically check for:
```
Comment (mrdewitt) [inline README.md]: Seems to me that rewording this without the numbered list would be clearer.  "It may be feasible to periodically check for OCSP notification that a package's certificate has been revoked.  Other channels for notifying about vulnerable packages have not yet been discussed/designed."

File: README.md [hunk 4]
```diff
@@ -8,9 +8,26 @@ Some new use cases for Web technology have motivated thinking about a multi-reso
...
 
+Physical web beacons may be located in a place that has no connectivity, and
+their owner may only visit to update them as often as their battery needs
+replacement: annually or less often. This leaves a large window during which a
+certificate could be compromised or a package could get out of date. We think
+there's no way to prevent a client from trusting its initial download of a
+package signed by a compromised certificate. When the client gets back to a
+public network, it should attempt to validate both the certificate and the
+package, likely by fetching the package's `Content-Location`.
```
Comment (mrdewitt) [inline README.md]: This makes me think that analogue to the HEAD request where the client could fetch only the manifest/index and revalidate stale portions would be useful.

File: README.md [hunk 5]
```diff
@@ -183,7 +203,12 @@ The example contains an HTML page and an image. The package is signed by the exa
 
 Important notes:
 
-1. The very first header in Package Header section of the package is **Package-Signature**, a new header that contains a signed hash of the Package Header section (not including Package-Signature header) and Content Index. It also contains a reference (via urn:uuid: UUID-based URL) to the part that contains the public key certificate (or if needed, a chain of certificates to the root CA).
+1. The very first header in Package Header section of the package is
```
Comment (mrdewitt) [inline README.md]: s/is/must be/

File: README.md (hunk 2 above)
Comment (jyasskin) [inline README.md]: Indeed, that was worded badly. I meant that the use case happens where global connectivity is expensive, and have now updated the PR wording.

File: README.md (hunk 3 above)
Comment (jyasskin) [inline README.md]: Done.

File: README.md (hunk 4 above)
Comment (jyasskin) [inline README.md]: Yeah. We should define an optimized update mechanism, although validation might be simpler.

File: README.md (hunk 5 above)
Comment (jyasskin) [inline README.md]: I'm going to undo the change to this section until we can update all the use cases.

File: README.md (hunk 1 above)
Comment (jyasskin) [inline README.md]: #44 takes agl's advice to use the same signing format as TLS (and I used a different context string to distinguish package signatures from handshake signatures). I'd like to postpone the question of exactly what origin-trusted certificates will convince browsers to load a package until a later PR.

The current format even looks like it'll represent packages that have only signatures that aren't origin-trusted. Although web browsers would refuse to load those packages, they're useful for Node and Publishers.

**Comment by dimich-g** [comment]
Points by @mrdewitt addressed, merged.
